be/src/exec/pipeline/exchange/mem_limited_chunk_queue.cpp

```diff
@@ -0,0 +1,528 @@
+
+        // 1. find block to flush
+        while (_next_flush_block) {
+            // don't flush the last block
+            if (_next_flush_block->next == nullptr) {
+                break;
+            }
+            if (_next_flush_block->block) {
+                // already flushed, should skip it
+                _next_flush_block = _next_flush_block->next;
+            } else {
+                // check if need flush
+                chunks.clear();
+                std::for_each(_next_flush_block->cells.begin(), _next_flush_block->cells.end(),
+                              [&chunks, this](Cell& cell) {
+                                  if (cell.chunk == nullptr || cell.used_count == _consumer_number) {
+                                      return;
+                                  }
+                                  cell.flushed = true;
+                                  chunks.emplace_back(cell.chunk);
+                              });
+                if (!chunks.empty()) {
+                    break;
+                }
+                _next_flush_block = _next_flush_block->next;
+            }
+        }
+        block = _next_flush_block;
+        DCHECK(block != nullptr) << "block can't be null";
+        if (block->next == nullptr) {
+            return Status::OK();
+        }
+        if (block->block != nullptr) {
+            return Status::OK();
+        }
+    }
+
+    DCHECK(!chunks.empty());
+
+    size_t max_serialize_size = 0;
+    size_t flushed_rows = 0, flushed_bytes = 0;
+    for (auto& chunk : chunks) {
+        for (const auto& column : chunk->columns()) {
+            max_serialize_size += serde::ColumnArraySerde::max_serialized_size(*column, _opts.encode_level);
+        }
+        flushed_rows += chunk->num_rows();
+        flushed_bytes += chunk->memory_usage();
+    }
+    std::shared_ptr<spill::Block> spill_block;
+
+    TEST_SYNC_POINT_CALLBACK("MemLimitedChunkQueue::after_calculate_max_serialize_size", &max_serialize_size);
+    raw::RawString serialize_buffer;
+    serialize_buffer.resize(max_serialize_size);
+    uint8_t* buf = reinterpret_cast<uint8_t*>(serialize_buffer.data());
+    uint8_t* begin = buf;
+    // 2. serialize data
+    for (auto& chunk : chunks) {
+        for (const auto& column : chunk->columns()) {
+            buf = serde::ColumnArraySerde::serialize(*column, buf, false, _opts.encode_level);
+            RETURN_IF(buf == nullptr, Status::InternalError("serialize data error"));
+        }
+    }
+    size_t content_length = buf - begin;
+    // 3. acquire block
+    spill::AcquireBlockOptions options;
+    options.block_size = content_length;
+    options.direct_io = false;
+    options.query_id = _state->query_id();
+    options.fragment_instance_id = _state->fragment_instance_id();
+    options.name = "mcast_local_exchange";
+    options.plan_node_id = _opts.plan_node_id;
+    ASSIGN_OR_RETURN(spill_block, _opts.block_manager->acquire_block(options));
```

Comment:
we shouldn't direct using BlockManager

Reply:
can we write it by spiller?

Reply:
> can we write it by spiller or add a stream interface?

It's not easy. The usage scenario of this queue is too different from that of spiller. I think it would be more appropriate to treat it as a stream interface for a spiller?